proposals/2025-01-02_direct-io.md

```diff
@@ -0,0 +1,166 @@
+# Direct I/O
+
+* **Owners:**
+  * [@machine424](https://github.com/machine424)
+
+* **Implementation Status:** Partially implemented
+
+* **Related Issues and PRs:**
+  * [Prometheus PR #15365](https://github.com/prometheus/prometheus/pull/15365)
+
+* **Other Docs or Links:**
+  * [Slack Discussion](https://cloud-native.slack.com/archives/C01AUBA4PFE/p1726674665380109)
+
+> This effort aims to experiment with direct I/O to determine whether it can enhance user
+experience and provide performance improvements.
+
+## Why
+
+Probably due to a lack of [understanding of good taste](https://yarchive.net/comp/linux/o_direct.html).
+
+The motivation behind this effort is to address the confusion surrounding page cache behavior,
+particularly in environments where various memory-related statistics and metrics are available; for
+instance, see [this issue](https://github.com/kubernetes/kubernetes/issues/43916) regarding
+containerized environments.
+
+This initiative was prompted by [@bboreham](https://github.com/bboreham)
+(refer to [the Slack thread](https://cloud-native.slack.com/archives/C01AUBA4PFE/p1726674665380109)).
+The effort aims to prevent misunderstandings and concerns regarding increased page cache usage, which
+might lead users or admins to make poor decisions, such as allocating more memory than necessary.
+Moreover, bypassing the cache when it is not required can help eliminate the overhead of data copying
+and reduce additional kernel overhead from reclaiming the cache. Thus, any performance improvements
+achieved through direct I/O are welcome.
+
+### Pitfalls of the Current Solution
+
+In addition to the issues mentioned above, sometimes the page cache generated during writes is not
+used at all. For instance, once chunks are written during compaction, they are opened via
+`mmap`, which renders the page cache produced during writing redundant and useless.
```

Comment:
I'm not sure that's entirely true, Linux has a merged page + buffer cache: https://www.thomas-krenn.com/en/wiki/Linux_Page_Cache_Basics#Page_Cache_or_Buffer_Cache

(The exception being OpenBSD, see https://github.com/prometheus/prometheus/issues/8799)

Reply:
You're right, I'll make that clearer.